lwesp/src/system/lwesp_sys_threadx.c

```diff
@@ -0,0 +1,235 @@
+    ULONG start = tx_time_get();
+    return tx_semaphore_get(p, !timeout ? TX_WAIT_FOREVER : LWESP_MS_TO_TICKS(timeout)) == TX_SUCCESS ? (tx_time_get() - start) * LWESP_TICK_PERIOD : LWESP_SYS_TIMEOUT;
+}
+
+uint8_t
+lwesp_sys_sem_release(lwesp_sys_sem_t* p) {
+    return tx_semaphore_put(p) == TX_SUCCESS ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_sem_isvalid(lwesp_sys_sem_t* p) {
+    return p->tx_semaphore_id != TX_CLEAR_ID ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_sem_invalid(lwesp_sys_sem_t* p) {
+    /* No need actions since all invalid are following delete, and delete make sure it is invalid */
+    return 1;
+}
+
+uint8_t
+lwesp_sys_mbox_create(lwesp_sys_mbox_t* b, size_t size) {
+
+    uint8_t rt = 0;
+    ULONG queue_total_size = size * sizeof(void *);
+
+    void *queue_mem = lwesp_mem_malloc(queue_total_size);
+    if (queue_mem != NULL) {
+        if (tx_queue_create(b, TX_NULL, sizeof(void *) / sizeof(ULONG), queue_mem, queue_total_size) == TX_SUCCESS) {
+            rt = 1;
+        } else {
+            lwesp_mem_free(queue_mem);
+        }
+    }
+
+    return rt;
+}
+
+uint8_t
+lwesp_sys_mbox_delete(lwesp_sys_mbox_t* b) {
+
+    (VOID)tx_queue_delete(b);
+    lwesp_mem_free(b->tx_queue_start);
+    return 1;
+}
+
+uint32_t
+lwesp_sys_mbox_put(lwesp_sys_mbox_t* b, void* m) {
+
+    ULONG start = tx_time_get();
+    (VOID)tx_queue_send(b, &m, TX_WAIT_FOREVER);
+    return tx_time_get() - start;
+}
+
+uint32_t
+lwesp_sys_mbox_get(lwesp_sys_mbox_t* b, void** m, uint32_t timeout) {
+
+    ULONG start = tx_time_get();
+    return tx_queue_receive(b, m, !timeout ? TX_WAIT_FOREVER : LWESP_MS_TO_TICKS(timeout)) == TX_SUCCESS ? (tx_time_get() - start) * LWESP_TICK_PERIOD : LWESP_SYS_TIMEOUT;
+}
+
+uint8_t
+lwesp_sys_mbox_putnow(lwesp_sys_mbox_t* b, void* m) {
+    return tx_queue_send(b, &m, TX_NO_WAIT) == TX_SUCCESS ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_mbox_getnow(lwesp_sys_mbox_t* b, void** m) {
+    return tx_queue_receive(b, m, TX_NO_WAIT) == TX_SUCCESS ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_mbox_isvalid(lwesp_sys_mbox_t* b) {
+    return b->tx_queue_id != TX_CLEAR_ID ? 1 : 0;
+}
+
+uint8_t
+lwesp_sys_mbox_invalid(lwesp_sys_mbox_t* b) {
+    /* No need actions since all invalid are following delete, and delete make sure it is invalid */
+    return 1;
+}
+
+uint8_t
+lwesp_sys_thread_create(lwesp_sys_thread_t* t, const char* name, lwesp_sys_thread_fn thread_func, void* const arg, size_t stack_size, lwesp_sys_thread_prio_t prio) {
+
+typedef VOID (*threadx_entry_t)(ULONG);
+    uint8_t rt = 0;
+
+    void *stack_mem = lwesp_mem_malloc(stack_size);
+    if (stack_mem != NULL) {
+        if (tx_thread_create(t, (CHAR *)name, (VOID (*)(ULONG))(thread_func), (ULONG)arg, stack_mem, stack_size, prio, prio, TX_NO_TIME_SLICE, TX_AUTO_START) == TX_SUCCESS) {
+            rt = 1;
+        } else {
+            lwesp_mem_free(stack_mem);
+        }
+    }
+
+    return rt;
+}
+
+uint8_t
+lwesp_sys_thread_terminate(lwesp_sys_thread_t* t) {
+
```

Comment:
If parameter is NULL, normally you should be able to kill yourself

Reply:
I should use tx_thread_terminate instead of tx_thread_delete. Should I delete the task and free stack memory? It is not possible for a thread to delete itself in ThreadX. So when a NULL is passed, there are problems. (only terminate, not delete)

Reply:
Here is the exact point. If you terminate yourself, there must be a way to free stack too. Can you even do this from itself thread?

Better to ask, how to properly terminate "itself" in ThreadX if you use dynamic stack allocation? Any callback that can be used when task gets terminated for such cleanup operation?

Reply:
Double checked with RTOS dev. Unless you create another thread/timer to do this (FreeRTOS is doing in idle thread while ThreadX does not have IDLE thread by default). 

I also go to cmsis_os wrapper for threadx for reference. It looks like they are ignoring this.
https://github.com/STMicroelectronics/x-cube-azrtos-h7/blob/HEAD/Middlewares/ST/cmsis_rtos_threadx/cmsis_os2.c.

It looks like this is not used in (deleting itself) the library but in some sample code project. Can I do termination only?

Reply:
Yes, you are right here. An idle thread would need to be implemented and then every thread to clean itself should notify idle thread and go to terminate state afterwards.

For the moment, please implement `return 0` if feature does not really work out.

Reply:
It is not required for CORE itself, it is instead used in some cases to demonstrate some applications. For example netconn client where one thread is used for demo and terminated after client connection finishes with server.

I will merge to develop branch.